enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,511 @@
+as part of an OLM based install at the time of this writing is immature, at least for some reviewers.
+- However, other reviewers wonder if we are in fact close from having either OLM or CVO operators being able to require
+an OLM operator to be present.
+- Using OLM avoids further increasing the install payload size.
+- Using OLM avoids unexpected hiccups in the feature's installation impacting initial cluster install or cluster upgrade.
+
+ #### CSI starting points
+ 
+ **WE ARE GOING WITH EPHEMERAL MODE**
+ 
+ **NO SecretProvideClass IS NEEDED**
+ 
+ With those simplifying decisions, a CSI driver with those characteristics still meets our needs.  And the storage 
+ SMEs recommend that approach to us.
+ 
+ Crafting the CSI driver in this way also helps us deal with SELinux.  Specifically (thanks to @jsafrane for details):
+ - when CRI-O starts any container, it updates the SELinux labels on all volumes for that container, where those labels
+ are unique to the container
+ - in the context of k8s, multiple containers within a pod have the same filesystem access
+ - so, only that specific pod can read files on that volume
+ - however, a file can only have one SELinux label
+ - so two or more pods cannot access the same file
+ - but if each pod has its own file, its own copy, of the shared data, things work easily from a SELinux 
+ perspective
+ - this is where using ephemeral CSI storage comes in
+ - using ephemeral based storage allows for use of a tmpfs (virtual memory filesystem) for storage
+ - tmpfs is fast and has negligible overhead
+ - it will have to reconstructed if the Pod moves
+ - but how it eases dealing with SELinux protection is the bonus we are taking
+ - so each pod gets its own tmpfs
+ - the kubelet leverages tmpfs in a similar fashion already
+
+#### End to end flow (from creation, to opt-in, to validation, and then injections)
+
+An admin creates a new cluster level custom resource for encapsulating the sharable `Secret` or `ConfigMap`.
+
+
+```yaml
+kind: ProjectedResource
+apiVersion: projected-resource.storage.openshift.io/v1
+metadata:
+  name: shared-cool-secret
+spec:
+  backingResource:
+    kind: Secret
+    apiVersion: v1
+    namespace: ns-one
+    name: cool-secret
+  description: "used to access my internal git repo, RHEL entitlements, or something else cool"
+status:
+  conditions:
+  ...
+```
+
+And sets up ACL/RBAC to control who can list/get such objects.
+
+```yaml
+rules:
+- apiGroups:
+  - projected-resource.storage.openshift.io
+  resources:
+  - projectedResources
+  resourceNames:
+  - shared-cool-secret
+  verbs:
+  - get
+  - list
+  - watch
+
+```
+
+A user queries this new type to see what is available to her/him.
+
+Then, to opt in, a user created Pod can cite a CSI volume references (see below for a more detailed explanation of them) that 
+references our new type:
+
+```yaml
+csi:
```

Comment:
IMO, whole "tree" would be nice. `csi` itself does not say where it belongs.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: some-pod
spec:
  containers:
    ...
  volumes:
      - name: secret
        csi:
          driver: projectedresources.storage.openshift.io
          volumeAttributes:
            projectedResourceName: shared-cool-secret
```

Reply:
I have the Pod level yaml a little bit further down but sure I can replicate that from the get go